src/test/fuzz/util.h

```diff
@@ -534,4 +544,121 @@ void ReadFromStream(FuzzedDataProvider& fuzzed_data_provider, Stream& stream) no
+    {
+    }
+
+    ~FuzzedSock() override
+    {
+    }
+
+    SOCKET Get() const override
+    {
+        assert(false && "Not implemented yet.");
+    }
+
+    SOCKET Release() override
+    {
+        assert(false && "Not implemented yet.");
+    }
+
+    void Reset() override
+    {
+        assert(false && "Not implemented yet.");
+    }
+
+    ssize_t Send(const void* data, size_t len, int flags) const override
+    {
+        constexpr std::array send_errnos{
+            EACCES,
+            EAGAIN,
+            EALREADY,
+            EBADF,
+            ECONNRESET,
+            EDESTADDRREQ,
+            EFAULT,
+            EINTR,
+            EINVAL,
+            EISCONN,
+            EMSGSIZE,
+            ENOBUFS,
+            ENOMEM,
+            ENOTCONN,
+            ENOTSOCK,
+            EOPNOTSUPP,
+            EPIPE,
+            EWOULDBLOCK,
+        };
+        if (m_fuzzed_data_provider.ConsumeBool()) {
+            return len;
+        }
+        const ssize_t r = m_fuzzed_data_provider.ConsumeIntegralInRange<ssize_t>(-1, len);
+        if (r == -1) {
+            SetFuzzedErrNo(m_fuzzed_data_provider, send_errnos);
+        }
+        return r;
+    }
+
+    ssize_t Recv(void* buf, size_t len, int flags) const override
+    {
+        constexpr std::array recv_errnos{
+            EAGAIN,
+            EBADF,
+            ECONNREFUSED,
+            EFAULT,
+            EINTR,
+            EINVAL,
+            ENOMEM,
+            ENOTCONN,
+            ENOTSOCK,
+            EWOULDBLOCK,
+        };
+        assert(buf != nullptr || len == 0);
+        if (len == 0 || m_fuzzed_data_provider.ConsumeBool()) {
+            const ssize_t r = m_fuzzed_data_provider.ConsumeBool() ? 0 : -1;
+            if (r == -1) {
+                SetFuzzedErrNo(m_fuzzed_data_provider, recv_errnos);
+            }
+            return r;
+        }
+        const std::vector<uint8_t> random_bytes = m_fuzzed_data_provider.ConsumeBytes<uint8_t>(
+            m_fuzzed_data_provider.ConsumeBool() ?
+                len :
+                m_fuzzed_data_provider.ConsumeIntegralInRange<size_t>(0, len));
+        if (random_bytes.empty()) {
+            const ssize_t r = m_fuzzed_data_provider.ConsumeBool() ? 0 : -1;
+            if (r == -1) {
+                SetFuzzedErrNo(m_fuzzed_data_provider, recv_errnos);
+            }
+            return r;
+        }
+        std::memcpy(buf, random_bytes.data(), random_bytes.size());
+        if (m_fuzzed_data_provider.ConsumeBool()) {
+            if (len > random_bytes.size()) {
+                std::memset((char*)buf + random_bytes.size(), 0, len - random_bytes.size());
+            }
+            return len;
+        }
+        if (m_fuzzed_data_provider.ConsumeBool() && std::getenv("FUZZED_SOCKET_FAKE_LATENCY") != nullptr) {
+            std::this_thread::sleep_for(std::chrono::milliseconds{2});
```

Comment:
`SOCKS5_RECV_TIMEOUT` can be lowered to 1 second for the tests or be the default 20 seconds.
In half of the cases this will sleep for 2 ms. So we need 500 `Recv()` calls where this sleep is executed in order to trigger a higher level timeout of 1 second.

I think this timeout will practically never be reached. Maybe change this to a few 100s of ms to make a timeout more likely?

Reply:
You are right - it can be lowered to 1ms or be the default 20sec. If we want to always produce a timeout, maybe better to sleep for `SOCKS5_RECV_TIMEOUT` instead of `std::chrono::milliseconds{2}`?

Why would it be fake sleeping? My understanding is that it will actually sleep in the middle of the fuzz test.

Reply:
Unfortunately the sleeping required to make the timeout trigger is not mockable: that's the reason for making the ugly 2 ms sleep opt-in via the environment variable `FUZZED_SOCKET_FAKE_LATENCY`. That environment is not assumed to be set except in special circumstances such as when targeted fuzzing is performed by security researchers and others who are willing to pay the 2 ms cost in order to get full coverage including the timeout logic.

Making the sleeping required to trigger the timeout mockable is likely out of scope for this PR (but would be nice to have!), so if the opt-in-to-timeout-via-`FUZZED_SOCKET_FAKE_LATENCY` workaround is deemed too ugly I could simply drop that commit from this PR, and we'll get partial instead of full coverage :)

Let me know what you think and I'll adjust accordingly :)